File: tools/record-thread-rows.mjs
#!/usr/bin/env node
// Record the TypeScript rows in suites/harness-thread-rows/cases.json, by running
// prism-harness-ts's real assistantRow(), toolResultRow() and threadView() -- the
// builders AgentRuntime records with and the view it replays -- on each case.
//
// The case file is edited as TEXT, replacing each `"ts": ...` string in place and
// recomputing `agrees`, so the recorder cannot reformat or retype the rows it
// records.
//
//   PRISM_HARNESS_TS=../prism-harness-ts node tools/record-thread-rows.mjs [--check]

import { readFileSync, writeFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const root = dirname(dirname(fileURLToPath(import.meta.url)));
const harnessRoot = process.env.PRISM_HARNESS_TS ?? join(root, '..', 'prism-harness-ts');
const harness = await import(pathToFileURL(join(harnessRoot, 'dist', 'index.js')).href);

const check = process.argv.includes('--check');
const path = join(root, 'suites', 'harness-thread-rows', 'cases.json');
let raw = readFileSync(path, 'utf8');
const document = JSON.parse(raw);

// The same conversion prism-harness-ts/test/thread-rows-corpus.test.ts makes.
export function rowsFor(testCase, { assistantRow, toolResultRow, threadView }) {
  if (testCase.fold !== undefined) return threadView(testCase.fold);

  const input = testCase.input;

  if (input.write === 'assistant') {
    return [
      assistantRow(
        input.content,
        input.tool_calls.map((call) => ({
          id: call.id,
          name: call.name,
          arguments: call.arguments,
          resultId: call.result_id ?? null,
          reasoningId: call.reasoning_id ?? null,
          reasoningSummary: call.reasoning_summary ?? null,
        })),
        input.additional_content,
        input.approval_requests,
      ),
    ];
  }

  return [
    toolResultRow(
      input.results.map((result) => ({
        tool_call_id: result.tool_call_id,
        tool_name: result.tool_name,
        args: result.args,
        result: result.result,
        tool_call_result_id: result.tool_call_result_id ?? null,
        artifacts: [],
      })),
      input.decisions,
    ),
  ];
}

/** The offset just past the JSON string literal that starts at `quote`. */
function stringEnd(text, quote) {
  for (let i = quote + 1; i < text.length; i += 1) {
    if (text[i] === '\\') i += 1;
    else if (text[i] === '"') return i + 1;
  }

  throw new Error('Unterminated string in the case file.');
}

const stale = [];

for (const testCase of document.cases) {
  const produced = JSON.stringify(rowsFor(testCase, harness));

  if (testCase.rows.ts !== produced) stale.push(testCase.id);

  const at = raw.indexOf(`"id": "${testCase.id}"`);
  const quote = raw.indexOf('"ts": ', at) + '"ts": '.length;
  raw = raw.slice(0, quote) + JSON.stringify(produced) + raw.slice(stringEnd(raw, quote));

  const agrees = produced === testCase.rows.php && produced === testCase.rows.py;
  const valueAt = raw.indexOf('"agrees": ', at) + '"agrees": '.length;
  const valueEnd = raw.startsWith('true', valueAt) ? valueAt + 4 : valueAt + 5;
  raw = raw.slice(0, valueAt) + String(agrees) + raw.slice(valueEnd);
}

if (check) {
  if (stale.length > 0) {
    process.stderr.write(`Stale ts rows: ${stale.join(', ')}\n`);
    process.exit(1);
  }

  process.stderr.write('TypeScript rows current.\n');
  process.exit(0);
}

writeFileSync(path, raw);
process.stderr.write(`Wrote ${document.cases.length} ts row set(s); ${stale.length} changed.\n`);

File: tools/record_thread_rows.py
#!/usr/bin/env python3
"""Record the Python rows in suites/harness-thread-rows/cases.json.

Runs prism-harness-py's real ``assistant_row()``, ``tool_result_row()`` and
``thread_view()`` -- the builders AgentRuntime records with and the view it
replays -- on each case.

The case file is edited as TEXT, replacing each ``"py": ...`` string in place and
recomputing ``agrees``, so the recorder cannot reformat or retype the rows it
records.

    PRISM_HARNESS_PY=../prism-harness-py python tools/record_thread_rows.py [--check]
"""

from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parent.parent
HARNESS = Path(os.environ.get("PRISM_HARNESS_PY", ROOT.parent / "prism-harness-py"))
sys.path.insert(0, str(HARNESS / "src"))

from prism_harness import (  # noqa: E402
    ToolCallInput,
    assistant_row,
    thread_view,
    tool_result_row,
)


def rows_for(case: dict[str, Any]) -> list[dict[str, Any]]:
    """The same conversion prism-harness-py/tests/test_thread_rows_corpus.py makes."""
    if "fold" in case:
        return thread_view(case["fold"])

    spec = case["input"]

    if spec["write"] == "assistant":
        return [
            assistant_row(
                spec["content"],
                [
                    ToolCallInput(
                        id=call["id"],
                        name=call["name"],
                        arguments=call["arguments"],
                        result_id=call.get("result_id"),
                        reasoning_id=call.get("reasoning_id"),
                        reasoning_summary=call.get("reasoning_summary"),
                    )
                    for call in spec["tool_calls"]
                ],
                spec["additional_content"],
                spec["approval_requests"],
            )
        ]

    return [
        tool_result_row(
            [
                {
                    "tool_call_id": result["tool_call_id"],
                    "tool_name": result["tool_name"],
                    "args": result["args"],
                    "result": result["result"],
                    "tool_call_result_id": result.get("tool_call_result_id"),
                    "artifacts": [],
                }
                for result in spec["results"]
            ],
            spec["decisions"],
        )
    ]


def encode(value: Any) -> str:
    """Compact JSON with non-ASCII and slashes left as they are, as the reference writes it."""
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def string_end(text: str, quote: int) -> int:
    i = quote + 1
    while i < len(text):
        if text[i] == "\\":
            i += 2
            continue
        if text[i] == '"':
            return i + 1
        i += 1
    raise ValueError("Unterminated string in the case file.")


def main() -> int:
    check = "--check" in sys.argv
    path = ROOT / "suites" / "harness-thread-rows" / "cases.json"
    raw = path.read_text(encoding="utf-8")
    document = json.loads(raw)
    stale: list[str] = []

    for case in document["cases"]:
        produced = encode(rows_for(case))

        if case["rows"]["py"] != produced:
            stale.append(case["id"])

        at = raw.index(f'"id": "{case["id"]}"')
        quote = raw.index('"py": ', at) + len('"py": ')
        raw = raw[:quote] + encode(produced) + raw[string_end(raw, quote) :]

        agrees = produced == case["rows"]["php"] and produced == case["rows"]["ts"]
        value_at = raw.index('"agrees": ', at) + len('"agrees": ')
        value_end = value_at + (4 if raw.startswith("true", value_at) else 5)
        raw = raw[:value_at] + ("true" if agrees else "false") + raw[value_end:]

    if check:
        if stale:
            print(f"Stale py rows: {', '.join(stale)}", file=sys.stderr)
            return 1
        print("Python rows current.", file=sys.stderr)
        return 0

    path.write_text(raw, encoding="utf-8", newline="")
    print(f"Wrote {len(document['cases'])} py row set(s); {len(stale)} changed.", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
